test/topology_custom/test_raft_recovery_user_data.py

```diff
@@ -0,0 +1,153 @@
+
+    if remove_dead_nodes_with == "remove":
+        # We must mark dead nodes as permanently dead so that they are ignored in topology commands. Without this step,
+        # ALTER KEYSPACE below would fail on the global token metadata barrier.
+        # For now, we do not have a specific API to mark nodes as dead so we use a work-around.
+        logging.info(f'Marking {dead_servers} as permanently dead')
+        await manager.remove_node(live_servers[0].server_id, dead_servers[0].server_id,
+                                  [dead_srv.ip_addr for dead_srv in dead_servers[1:]],
+                                  expected_error='Removenode failed')
+
+        logging.info(f'Desceasing RF of {ks_name} to 0 in dc2')
+        for i in range(1, rf + 1):
+            # ALTER KEYSPACE with tablets can decrease RF only by one.
+            await cql.run_async(f"""ALTER KEYSPACE {ks_name} WITH replication =
+                                    {{'class': 'NetworkTopologyStrategy', 'dc1': {rf}, 'dc2': {rf - i}}}""")
+
+        logging.info(f'Removing {dead_servers}')
+        for i, being_removed in enumerate(dead_servers):
+            ignored = [dead_srv.ip_addr for dead_srv in dead_servers[i + 1:]]
+            initiator = live_servers[i]
+            await manager.remove_node(initiator.server_id, being_removed.server_id, ignored)
+    else:
+        logging.info(f'Replacing {dead_servers}')
+        for i, being_replaced in enumerate(dead_servers):
+            replace_cfg = ReplaceConfig(replaced_id=being_replaced.server_id, reuse_ip_addr=False, use_host_id=True,
+                                        ignore_dead_nodes=[dead_srv.ip_addr for dead_srv in dead_servers[i + 1:]])
+            await manager.server_add(replace_cfg=replace_cfg, property_file=property_file_dc2)
+
+    logging.info(f'Unsetting the recovery_leader config on {live_servers}')
+    for srv in live_servers:
+        await manager.server_update_config(srv.server_id, 'recovery_leader', '')
+
+    logging.info(f'Deleting persistent data of group 0 {first_group0_id} on {live_servers}')
+    for h in hosts:
+        await delete_raft_group_data(first_group0_id, cql, h)
+
+    logging.info('Performing consistency checks after the recovery procedure')
+    await wait_for_cdc_generations_publishing(cql, hosts, time.time() + 60)
+    await check_token_ring_and_group0_consistency(manager)
+    await check_system_topology_and_cdc_generations_v3_consistency(manager, hosts)
+
+    # Disable load balancer on the topology coordinator node so that an ongoing tablet migration doesn't fail the
+    # check_system_topology_and_cdc_generations_v3_consistency call below. global_tablet_token_metadata_barrier can
+    # suddenly make fence_version inconsistent among nodes.
+    await manager.api.disable_tablet_balancing(live_servers[0].ip_addr)
+
+    logger.info('Adding a new server to dc2')
+    new_server = await manager.server_add(config=cfg, property_file=property_file_dc2)
```

Comment:
perhaps we could also bring back some RF to this DC afterwards

Reply:
Good idea.

The replace case has RF unchanged, so we only must change the removenode case. I would add 3 nodes instead of 1 and bring back RF to 2. This will basically test moving dc back to its previous state, which I'll like. But I'll have to check this test's running time.